Add single and multi branch cicd for awscodebuild (#1032)
astronomer/docs: astro/ci-cd.md

To automate code deploys to a single Deployment using [AWS CodeBuild](https://aws.amazon.com/codebuild/), complete the following setup in a Git-based repository hosting an Astro project:

1. In your AWS CodeBuild pipeline configuration, add the following environment variables:

    - `ASTRONOMER_KEY_ID`: Your Deployment API key ID
    - `ASTRONOMER_KEY_SECRET`: Your Deployment API key secret
    - `ASTRONOMER_DEPLOYMENT_ID`: The Deployment ID of your production deployment

    Be sure to set the values for your API credentials as secret.

2. At the root of your Git repository, add a [buildspec.yml](https://docs.aws.amazon.com/codebuild/latest/userguide/build-spec-ref.html#build-spec-ref-example) file that includes the following script:

   ```yaml
   
   version: 0.2

   phases:
     install:
       runtime-versions:
         python: latest
   
     build:
       commands:
         - echo "${CODEBUILD_WEBHOOK_HEAD_REF}"
         - export ASTRONOMER_KEY_ID="${ASTRONOMER_KEY_ID}"
         - export ASTRONOMER_KEY_SECRET="${ASTRONOMER_KEY_SECRET}"
         - curl -sSL install.astronomer.io | sudo bash -s
         - astro deploy "${ASTRONOMER_DEPLOYMENT_ID}" -f
   

> 

Reviewer: ```suggestion
    ```
```

>     ```
>     When configuring the AWS CodeBuild project, create a webhook source event using the source provider where your Astro project is hosted (i.e. Bitbucket, GitHub, etc.). After doing so, configure the Primary source webhook events to trigger for event type `PUSH`. After setting those configurations, this buildspec.yml triggers a code push to an Astro Deployment every time a commit or pull request is merged to the `main` branch of your repository.
> </TabItem>

Reviewer: ```suggestion
    ```

    When configuring the AWS CodeBuild project, create a webhook source event using the source provider where your Astro project is hosted (i.e. Bitbucket, GitHub, etc.). After doing so, configure the Primary source webhook events to trigger for event type `PUSH`. After setting those configurations, this buildspec.yml triggers a code push to an Astro Deployment every time a commit or pull request is merged to the `main` branch of your repository.

</TabItem>
```

    - `DEV_DEPLOYMENT_ID`: The Deployment ID of your Development Deployment

2. At the root of your Git repository, add a [buildspec.yml](https://docs.aws.amazon.com/codebuild/latest/userguide/build-spec-ref.html#build-spec-ref-example) that includes the following script:

   ```yaml
   
   version: 0.2
   
   phases:
     install:
       runtime-versions:
         python: latest
   
     build:
       commands:
         - |
           if expr "${CODEBUILD_WEBHOOK_HEAD_REF}" : "refs/heads/main" >/dev/null; then
             export ASTRONOMER_KEY_ID="${PROD_ASTRONOMER_KEY_ID}"
             export ASTRONOMER_KEY_SECRET="${PROD_ASTRONOMER_KEY_SECRET}"
             curl -sSL install.astronomer.io | sudo bash -s
             astro deploy "${PROD_DEPLOYMENT_ID}" -f
           fi
         - |
           if expr "${CODEBUILD_WEBHOOK_HEAD_REF}" : "refs/heads/dev" >/dev/null; then
             export ASTRONOMER_KEY_ID="${DEV_ASTRONOMER_KEY_ID}"
             export ASTRONOMER_KEY_SECRET="${DEV_ASTRONOMER_KEY_SECRET}"
             curl -sSL install.astronomer.io | sudo bash -s
             astro deploy "${DEV_DEPLOYMENT_ID}" -f
           fi

>    

Reviewer: ```suggestion
    ```
```

</TabItem>
</Tabs>

### AWS CodeBuild

<Tabs
    defaultValue="awscodebuildstandard"
    values={[
        {label: 'Standard', value: 'awscodebuildstandard'},
        {label: 'Multi-branch', value: 'awscodebuildmultibranch'},
    ]}>
<TabItem value="awscodebuildstandard">

> To automate code deploys to a single Deployment using [AWS CodeBuild](https://aws.amazon.com/codebuild/), complete the following setup in a Git-based repository hosting an Astro project:

Reviewer: I'm unfamiliar with CodeBuild, is your code stored in AWS or is it just a build tool? How is the buildspec triggered from GitHub/BitBucket/etc?

Author: Hey @jwitz with AWS CodeBuild, your code can be stored in GitHub, Bitbucket, etc. AWS CodeBuild uses a [webhook event](https://docs.aws.amazon.com/codebuild/latest/userguide/webhooks.html) to recognize when pushes, merges, and other git actions happen in a repository to trigger the AWS CodeBuild pipeline. Jenkins behaves similarly to this as well.

1. In your AWS CodeBuild pipeline configuration, add the following environment variables:

    - `ASTRONOMER_KEY_ID`: Your Deployment API key ID
    - `ASTRONOMER_KEY_SECRET`: Your Deployment API key secret

>     - `ASTRONOMER_DEPLOYMENT_ID`: The Deployment ID of your production deployment

Reviewer: We need to specify `CODEBUILD_WEBHOOK_HEAD_REF` here and explain what it does.

Author: Hey @jwitz! The CODEBUILD_WEBHOOK_HEAD_REF is an environment variable that is made available by default in AWS CodeBuild. When you are using a webhook to trigger a CodeBuild job, this variable is set with the branch where the webhook action originated from. So you don't need to set it manually.

    Be sure to set the values for your API credentials as secret.

2. At the root of your Git repository, add a [buildspec.yml](https://docs.aws.amazon.com/codebuild/latest/userguide/build-spec-ref.html#build-spec-ref-example) file that includes the following script:

>    <pre><code parentName="pre">{`

Reviewer: ```suggestion
   ```
```

   
   version: 0.2

   phases:
     install:
       runtime-versions:
         python: latest
   
     build:
       commands:
         - echo "${CODEBUILD_WEBHOOK_HEAD_REF}"
         - export ASTRONOMER_KEY_ID="${ASTRONOMER_KEY_ID}"
         - export ASTRONOMER_KEY_SECRET="${ASTRONOMER_KEY_SECRET}"
         - curl -sSL install.astronomer.io | sudo bash -s
         - astro deploy "${ASTRONOMER_DEPLOYMENT_ID}" -f
   

>    `}</code></pre>

Reviewer: ```suggestion
   ```
```

Reviewer: This suggestion _should_ read as adding three tick marks (`). Thanks GitHub!

>     When configuring the AWS CodeBuild project, create a webhook source event using the source provider where your Astro project is hosted (i.e. Bitbucket, GitHub, etc.). After doing so, configure the Primary source webhook events to trigger for event type `PUSH`. After setting those configurations, this buildspec.yml triggers a code push to an Astro Deployment every time a commit or pull request is merged to the `main` branch of your repository.

Reviewer: ```suggestion
3. In your AWS CodeBuild project, create a [webhook event](https://docs.aws.amazon.com/codebuild/latest/userguide/webhooks.html) for the source provider where your Astro project is hosted, such as BitBucket or GitHub. When configuring the webhook, select an event type of `PUSH`.

Your buildspec.yml file now triggers a code push to an Astro Deployment every time a commit or pull request is merged to the `main` branch of your repository.
```
A few questions here:
- What is the "source provider"? Is that where our Git repo is hosted?
- I found a "Primary source webhook" for GitHub but not for BitBucket. Can we generalize these instructions?

Even better, can we include an example configuration that someone could run using the [AWS CLI](https://docs.aws.amazon.com/cli/latest/reference/codebuild/create-webhook.html)?

Author: Correct, source provider is where the Git repo is hosted (i.e. GitHub, BitBucket)! I think BitBucket, GitHub, and AWS CodeCommit are the only 3 options for Git source repo's in AWS CodeBuild.

Are you asking for a configuration that someone could use to set up this same AWS CodeBuild process via the AWS CLI? I'm sure that the AWS CLI commands exist to set this up, but that wouldn't be uniform with the rest of the CI/CD docs on this page and what we recommend as a starting point (See GitHub Actions, CircleCI, Jenkins, etc.).

</TabItem>

<TabItem value="awscodebuildmultibranch">

> To automate code deploys across multiple Deployments using [AWS CodeBuild](https://aws.amazon.com/codebuild/), complete the following setup in a Git-based repository hosting an Astro project:

Reviewer: ```suggestion
To automate code deploys across multiple Deployments using [AWS CodeBuild](https://aws.amazon.com/codebuild/), complete the following setup in a Git-based repository in CodeBuild hosting an Astro project:
```
Is this correct?

Author: I would say that this isn't correct. CodeBuild doesn't behave as a Git repository, but a CI/CD tool. Some Git repository services contain built-in CI/CD tools (i.e. GitHub Actions, GitLab, BitBucket), but there are other standalone CI/CD tools out there that do not act as Git repositories (i.e. Jenkins, CircleCI, AWS CodeBuild, etc.)

1. In your AWS CodeBuild pipeline configuration, add the following environment variables:

>     - `PROD_ASTRONOMER_KEY_ID`: Your Production Deployment API key ID

Reviewer: Same here, need to specify `CODEBUILD_WEBHOOK_HEAD_REF`

Author: Hey @jwitz! The CODEBUILD_WEBHOOK_HEAD_REF is an environment variable that is made available by default in AWS CodeBuild. When you are using a webhook to trigger a CodeBuild job, this variable is set with the branch where the webhook action originated from. So you don't need to set it manually.

Author: So basically, it helps the CI/CD process identify if the developer pushed commits to dev/staging/main/etc. and we can build conditional logic (based on the branch they are working in) to trigger a deploy to an environment associated to that branch.

    - `PROD_DEPLOYMENT_ID`: The Deployment ID of your Production Deployment
    - `DEV_ASTRONOMER_KEY_ID`: Your Development Deployment API key ID
    - `DEV_ASTRONOMER_KEY_SECRET`: Your Development Deployment API key secret
    - `DEV_DEPLOYMENT_ID`: The Deployment ID of your Development Deployment

2. At the root of your Git repository, add a [buildspec.yml](https://docs.aws.amazon.com/codebuild/latest/userguide/build-spec-ref.html#build-spec-ref-example) that includes the following script:

>    <pre><code parentName="pre">{`

Reviewer: ```suggestion
   ```yaml
```

   ```yaml
   
   version: 0.2
   
   phases:
     install:
       runtime-versions:
         python: latest
   
     build:
       commands:
         - |

>            if expr "${CODEBUILD_WEBHOOK_HEAD_REF}" : "refs/heads/main" >/dev/null; then

Reviewer: ```suggestion
           if expr "${CODEBUILD_WEBHOOK_HEAD_REF}" : "refs/heads/main" >/prod/null; then
```

Reviewer: Do we need a different webhook for prod/ dev?

Author: @jwitz 

No need! I have tested this script in action and both need to be `/dev/null` because that's the result of a BASH expression when two strings are equal.

             export ASTRONOMER_KEY_SECRET="${PROD_ASTRONOMER_KEY_SECRET}"
             curl -sSL install.astronomer.io | sudo bash -s
             astro deploy "${PROD_DEPLOYMENT_ID}" -f
           fi
         - |
           if expr "${CODEBUILD_WEBHOOK_HEAD_REF}" : "refs/heads/dev" >/dev/null; then
             export ASTRONOMER_KEY_ID="${DEV_ASTRONOMER_KEY_ID}"
             export ASTRONOMER_KEY_SECRET="${DEV_ASTRONOMER_KEY_SECRET}"
             curl -sSL install.astronomer.io | sudo bash -s
             astro deploy "${DEV_DEPLOYMENT_ID}" -f
           fi
   

>    `}</code></pre>

Reviewer: ```suggestion
   ```
```

Reviewer: This suggestion _should_ read as adding three tick marks (`). Thanks GitHub!

>    When configuring the AWS CodeBuild project, create a webhook source event using the source provider where your Astro project is hosted (i.e. Bitbucket, GitHub, etc.). After doing so, configure the Primary source webhook events to trigger for event type `PUSH`. After setting those configurations, this buildspec.yml triggers a code push to an Astro Deployment every time a commit or pull request is merged to the `dev` and `main` branches of your repository.

Reviewer: Same edits from the previous setup apply here.
